### PR TITLE
Make `HorovodPlugin.all_gather` return a tensor

### LINKED ISSUE #9695: Horovod all_gather function refactor
## Proposed refactoring or deprecation

Current Horovod training type plugin collective function all_gather() is calling horovod all_gather() and convert result to a list

1. Horovod support allgather_object which support return a list of tensor
https://horovod.readthedocs.io/en/stable/_modules/horovod/torch/functions.html#allgather_object

2. Revisit the use cases to see do we need to return a list of tensor here, as the training_type_plugin all_gather() api is defined to return a tensor now.
### Motivation

Have correct and consistent collective behavior

### Pitch
```
    def all_gather(
        self, result: Union[torch.Tensor], group: Optional[Any] = dist_group.WORLD, sync_grads: bool = False
    ) -> torch.Tensor:
        if group is not None and group != dist_group.WORLD:
            raise ValueError("Horovod does not support allgather using a subcommunicator at this time. Unset `group`.")

        if len(result.shape) == 0:
            # Convert scalars to single dimension tensors
            result = result.reshape(1)

        # sync and gather all
        self.join()
        gathered = hvd.allgather(result)
        gathered_result = list(gathered.split(1, dim=0))
        return gathered_result
```

[RFC]
Option 1
```
    def all_gather(
        self, result: Union[torch.Tensor], group: Optional[Any] = dist_group.WORLD, sync_grads: bool = False
    ) -> torch.Tensor:
        if group is not None and group != dist_group.WORLD:
            raise ValueError("Horovod does not support allgather using a subcommunicator at this time. Unset `group`.")

        if len(result.shape) == 0:
            # Convert scalars to single dimension tensors
            result = result.reshape(1)

        # sync and gather all
        self.join()
        return hvd. allgather_object(result)
```

Option 2
```
    def all_gather(
        self, result: Union[torch.Tensor], group: Optional[Any] = dist_group.WORLD, sync_grads: bool = False
    ) -> torch.Tensor:
        if group is not None and group != dist_group.WORLD:
            raise ValueError("Horovod does not support allgather using a subcommunicator at this time. Unset `group`.")

        if len(result.shape) == 0:
            # Convert scalars to single dimension tensors
            result = result.reshape(1)

        # sync and gather all
        self.join()
        return  hvd.allgather(result)t)
```

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #9695 

The Horovod Plugin is the only plugin whose `all_gather` returns a List[torch.Tensor] instead of a regular tensor. this is despite the support or allgather here: https://horovod.readthedocs.io/en/stable/_modules/horovod/torch/mpi_ops.html#allgather

This PR updates the implementation of all gather here such that:
- It's consistent with the typehint provided (returning a Tensor, not a list)
- It's consistent with the training type plugin API

I'm unclear why it was implemented like this before

This is part of https://github.com/PyTorchLightning/pytorch-lightning/issues/7534
and came up during review of https://github.com/PyTorchLightning/pytorch-lightning/pull/9414 and https://github.com/PyTorchLightning/pytorch-lightning/pull/9677/

Note: TrainingTypePlugin.all_gather is not called *anywhere* in the trainer. this is purely to power the LightningModule's all_gather's implementation. So this would be a breaking change iff someone was training with horovod and they depended on `LightningModule.all_gather` 

### Does your PR introduce any breaking changes? If yes, please list them.

Yes, this makes an in-place change to `HorovodPlugin.all_gather` to return a new return type

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/horovod.py [hunk 1]
```diff
@@ -183,9 +183,7 @@ def all_gather(
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/horovod.py:183]: Can you fix the type annotation for `result`?

**Review by carmocca — APPROVED** [review]
cc @tgaddair as the original author in https://github.com/PyTorchLightning/pytorch-lightning/pull/3775

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
